autode/opt/optimisers/trm.py

```diff
@@ -0,0 +1,461 @@
+        cart_coords = CartesianCoordinates(self._species.coordinates)
+        if self._coord_type == "cart":
+            self._coords = cart_coords
+
+        else:
+            primitives = self._primitives
+            if len(primitives) < cart_coords.expected_number_of_dof:
+                logger.info(
+                    "Had an incomplete set of primitives. Adding "
+                    "additional distances"
+                )
+                for i, j in combinations(range(self._species.n_atoms), 2):
+                    primitives.append(Distance(i, j))
+            self._coords = DIC.from_cartesian(
+                x=cart_coords, primitives=primitives
+            )
+
+        return None
+
+    def _step(self) -> None:
+
+        self._coords.h = self._updated_h()
+        self._update_trust_radius()
+
+        if self._damp_if_required():
+            logger.info("Skipping quasi-NR step after damping")
+            return None
+
+        rfo_h_eff = self._get_rfo_minimise_h_eff()
+        rfo_step = np.matmul(-np.linalg.inv(rfo_h_eff), self._coords.g)
+        rfo_step_size = self._get_cart_step_size_from_step(rfo_step)
+
+        if rfo_step_size < self.alpha:
+            logger.info("Taking a pure RFO step")
+            step = rfo_step
+        else:
+            try:
+                qa_h_eff = self._get_trm_minimise_h_eff()
+                qa_step = np.matmul(-np.linalg.inv(qa_h_eff), self._coords.g)
+                logger.info("Taking a TRM/QA step optimised to trust radius")
+                step = qa_step
+
+            except OptimiserStepError as exc:
+                logger.debug(f"TRM/QA step failed: {str(exc)}")
+                # if TRM/QA failed, take scaled RFO step
+                scaled_step = rfo_step * (self.alpha / rfo_step_size)
+
+                logger.info(
+                    "Taking a scaled RFO step approximately within "
+                    "trust radius"
+                )
+                step = scaled_step
+
+        self._coords.allow_unconverged_back_transform = False
+        self._coords = self._coords + step  # finally, take the step!
+        self._coords.allow_unconverged_back_transform = True
+
+        step_size = self._get_cart_step_size_from_step(step)
+        logger.info(f"Size of step taken (in Cartesian) = {step_size:.3f} Å")
+
+        return None
+
+    def _get_cart_step_size_from_step(self, step: np.ndarray) -> float:
+        """
+        Obtains the Cartesian step size given a step in the current
+        coordinate system (e.g., Delocalised Internal Coordinates)
+
+        Args:
+            step (np.ndarray): The step in the current coordinates
+
+        Returns:
+            (float): The step size in Cartesian
+        """
+        new_coords = self._coords + step
+        cart_delta = new_coords.to("cart") - self._coords.to("cart")
+        return float(np.linalg.norm(cart_delta))
+
+    def _get_rfo_minimise_h_eff(self) -> np.ndarray:
```

Comment:
autodE has a nice separation between coordinate stuff and optimiser stuff, and calculating what is essentially the rfo step seems like an optimiser thing? Also, properties would get evaluated by tab completion or inspect, so I think it may be better to have a method. (The reason I obtain the effective hessian instead of the step is that it is needed for GDIIS interpolation, which I want to implement later) Otherwise, there is no reason to get the hessian, instead of simply getting the step

Reply:
@t-young31 autodE has a nice separation between coordinate stuff and optimiser stuff, and calculating what is essentially the rfo step seems like an optimiser thing? Also, properties would get evaluated by tab completion or inspect, so I think it may be better to have a method. (The reason I obtain the effective hessian instead of the step is that it is needed for GDIIS interpolation, which I want to implement later) Otherwise, there is no reason to get the hessian, instead of simply getting the step

Reply:
yep. you're completely right 💯